dynamic/internal/shim/loader.go

```diff
@@ -0,0 +1,263 @@
+			return "", err
+		}
+		pulumiHome = filepath.Join(home, ".pulumi")
+	}
+
+	return filepath.Join(pulumiHome, "dynamic_tf_plugins"), nil
+}
+
+type provider struct {
+	tfprotov6.ProviderServer
+
+	name, version string
+
+	close func() error
+}
+
+func (p provider) Name() string { return p.name }
+
+func (p provider) Version() string { return p.version }
+
+func (p provider) Close() error { return p.close() }
+
+// ignore signals to the reader that a returned value is intentionally ignored.
+//
+// It's usage should always be accompanied by a "TO DO" or a explanation.
+func ignore[T any](t T) {}
+
+func getProviderServer(
+	ctx context.Context, addr tfaddr.Provider, version getproviders.VersionConstraints,
+) (Provider, error) {
+	cacheDir, err := getPluginCache()
+	if err != nil {
+		return nil, err
+	}
+
+	systemCache := providercache.NewDir(cacheDir)
+
+	// Look through existing packages to see if the package we want is already downloaded.
+	for k, packages := range systemCache.AllAvailablePackages() {
+		if k != addr {
+			continue
+		}
+		// packages is sorted by precedence, so the first cached result is safe to
+		// use.
+		acceptable := versions.MeetingConstraints(version)
+		for _, p := range packages {
+			if acceptable.Has(p.Version) {
+				slog.InfoContext(ctx, "Found cached provider",
+					slog.Any("addr", addr.String()),
+					slog.Any("version", p.Version.String()))
+				p := p
+				return runProvider(ctx, &p)
+			}
+		}
+	}
+
+	// We have not found a package that fits our constraints, so we need to download
+	// one.
+
+	source := getproviders.NewRegistrySource(disco.New())
+
+	availableVersions, warnings, err := source.AvailableVersions(ctx, addr)
+	if err != nil {
+		// TODO Handle error kinds with distinct error messages
+		return nil, err
+	}
+
+	// TODO: Don't ignore warnings
+	ignore(warnings)
+
+	desiredVersion := availableVersions.NewestInSet(versions.MeetingConstraints(version))
+	if desiredVersion == versions.Unspecified {
+		return nil, fmt.Errorf("Could not resolve a version from %s: %s", addr, version)
+	}
+
+	meta, err := source.PackageMeta(ctx, addr, desiredVersion, getproviders.CurrentPlatform)
+	if err != nil {
+		return nil, err
+	}
+
+	_, err = systemCache.InstallPackage(ctx, meta, nil)
+	if err != nil {
+		return nil, err
+	}
+
+	p := systemCache.ProviderVersion(addr, desiredVersion)
+	contract.Assertf(p != nil, "We just downloaded (%s,%s) so it should be in the cache", addr, desiredVersion)
+
+	return runProvider(ctx, p)
+}
+
+// runProvider produces a provider factory that runs up the executable
+// file in the given cache package and uses go-plugin to implement
+// providers.Interface against it.
+func runProvider(ctx context.Context, meta *providercache.CachedProvider) (Provider, error) {
+	execFile, err := meta.ExecutableFile()
+	if err != nil {
+		return nil, err
+	}
+
+	config := &plugin.ClientConfig{
+		HandshakeConfig:  tfplugin.Handshake,
+		Logger:           logging.NewProviderLogger(""),
```

Comment:
Is this correct? Is logging handled in syncStdout/err below?

Reply:
We don't need to set a prefix, so it looks correct to me. We might want to iterate on how logging works here later, but this seems to work for now.